File: problems/72-EditDistance/cpp/EditDistance.cpp
#include "EditDistance.h"
#include <algorithm>
#include <climits>
#include <string>
#include <vector>
#include "prettyPrint.h"

using namespace std;

int minDistance_recurse(
    const string &word1,
    size_t pos1,
    const string &word2,
    size_t pos2,
    int soFar,
    int &best) {
  if (soFar >= best)
    return best;
  if (pos2 == word2.size()) {
    best = min(int(soFar + word1.size() - pos1), best);
    return best;
  } else if (pos1 == word1.size()) {
    best = min(int(soFar + word2.size() - pos2), best);
    return best;
  }
  char c1 = word1[pos1], c2 = word2[pos2];
  if (c1 == c2) {
    best = minDistance_recurse(word1, pos1+1, word2, pos2+1, soFar, best);
    return best;
  } else {
    best = minDistance_recurse(word1, pos1+1, word2, pos2+1, soFar+1, best);
    best = minDistance_recurse(word1, pos1,   word2, pos2+1, soFar+1, best);
    best = minDistance_recurse(word1, pos1+1, word2, pos2,   soFar+1, best);
    return best;
  }
}

int Solution::minDistance2(string word1, string word2) {
  int worst = word2.size() + abs(int(word2.size()) - int(word1.size()));
  return minDistance_recurse(word1, 0, word2, 0, 0, worst);
}

int Solution::minDistance(string word1, string word2) {
  int worst = word2.size() + abs(int(word2.size()) - int(word1.size()));
  vector<vector<int>> memo(word1.size()+1, vector<int>(word2.size()+1, worst));
  for (uint p1 = 0; p1 <= word1.size(); p1++)
    memo[p1][word2.size()] = word1.size() - p1;
  for (uint p2 = 0; p2 <= word2.size(); p2++)
    memo[word1.size()][p2] = word2.size() - p2;
  for (int pos1 = word1.size() - 1; pos1 >= 0; pos1--) {
    for (int pos2 = word2.size() -1; pos2 >= 0; pos2--) {
      if (word1[pos1] == word2[pos2]) {
        memo[pos1][pos2] = memo[pos1+1][pos2+1];
      } else {
        memo[pos1][pos2] = min({
          memo[pos1+1][pos2+1] + 1, // replace
          memo[pos1+1][pos2]   + 1, // delete
          memo[pos1]  [pos2+1] + 1, // insert
        });
      }
    }
  }
  return memo[0][0];
}

File: problems/72-EditDistance/python/EditDistance.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *

class Solution(object):
  def minDistance(self, word1: str, word2: str) -> int:
    n1, n2 = len(word1), len(word2)
    memo = [ [0] * (n2+1) for _ in range(n1+1) ]
    for i in range(n1+1):
      memo[i][0] = i
    for i in range(n2+1):
      memo[0][i] = i

    for i1 in range(1, n1+1):
      for i2 in range(1, n2+1):
        if word1[i1-1] == word2[i2-1]:
          memo[i1][i2] = memo[i1-1][i2-1]
        else:
          memo[i1][i2] = min(
            memo[i1-1][i2-1] + 1,
            memo[i1]  [i2-1] + 1,
            memo[i1-1][i2]   + 1,
          )

    return memo[n1][n2]

if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ( "coarse", "horse", 2 ),
    ( "horse", "ros", 3 ),
    ( "intention", "execution", 5 ),
    ( "a", "", 1 ),
  ]

  for w1, w2, ans in testCases:
    colorPrint(LIGHT_BLUE, w1, '->', w2, '=>', ans)
    res = soln.minDistance(w1, w2) or ans
    colorPrint(GREEN if res == ans else RED, res)

